text/0098-generalized-templates.md

```diff
@@ -0,0 +1,277 @@
+
+For users of Cedar who do not use a schema, type declarations for generalized templates are optional. Templates will not be validated, but type checking will be performed during link-time. While this ensures that the provided values match the type of slot they are filling, it does not ensure that the policy won't experience a run-time type error.
+
+## Basic example
+
+### Example 1 
+
+Suppose you are building a document signing application. For security purposes, you want to make sure that whenever someone shares a document the receiver can only access it for x amount of days. We can use generalized templates to define this as shown below. 
+
+Everytime this template is instantiated the slots ```?principal``` and ```?resource``` will be replaced with the entity type that represents the person that you want to share the document with and the entity type representing the document that you want to share. Additionally ```?startTime``` and ```?endTime``` will be replaced with the datetime bounds.
+
+By using generalized templates compared to an attribute of an entity to express this policy, we do not have to store information that is only relevant to our authorization decisions within an external data store. This allows us to have a cleaner seperation of concerns. 
+
+#### Schema 
+```
+entity Document = {
+    owner: Person 
+};
+entity Person;
+
+action Access appliesTo {
+    principal: FS::Person,
+    resource: [FS::Document],
+    context: { date: datetime }
+};
+```
+
+#### Template 
+```
+template(?startTime: datetime, ?endTime: datetime) =>
+permit(principal == ?principal, 
+action, 
+resource == ?resource) when 
+{ ?startTime <= context.now && 
+context.now <= ?endTime };
+```
+
+### Example 2
+In this example, suppose we have a file sharing application where we can create folders and documents. Every document created must be in a folder. We want to express the policy that if you are an owner of the folder then you can access any document within it. We can use generalized templates to define this as shown below. 
+
+Now everytime we create a folder, we link ```?resource``` with that folder to express this relationship. 
+
+An alternative way to do this is to store an additional owner attribute for every document, however this has a couple of drawbacks compared to using Generalized Templates. 
+
+(1) We lose the intent behind why we are authoring this policy. 
+
+(2) When transferring ownership of a folder to someone else you would have to manually update the owner attribute for all document's that belong to that folder 
+
+#### Schema
+```
+namespace FS {
+    entity Folder;
+    entity Document in Folder;
+    entity Person = {
+        owned_folders: Set<Folder>
+    };
+}
+
+action AccessDocument appliesTo {
+    principal: FS::Person,
+    resource: FS::Document,
+};
+```
+
+#### Template
+```
+permit(principal, action == Action::"AccessDocument", resource in ?resource) when { 
```

Comment:
As others pointed out above, it feels like it'd be clearer to force the `template` syntax everywhere a template is used (even though there would not be other arguments), it'd make it clearer than this policy just being an "implicit" template by just using the special `?resource` and `?principal`. I understand why they are special, but I'm thinking if there's a simple way to make it more explicit.